### PR TITLE
Gather validator with beacon node plugin

### PR DESCRIPTION
**Pull request by mhchia**
### What was wrong?
Merge the rest of #465 back.

### How was it fixed?
Based on https://github.com/ethereum/trinity/pull/587.
- Parse genesis state and block, and privkeys from pre-generated files from https://github.com/ethereum/trinity/pull/517.
- Spin up `Validator`, `SlotTicker` in beacon node plugin based on the genesis files and privkeys.
- Add script `run_beacon_node.py` and tests for `trinity-beacon`.

### To-Do
- [x] Fix tests/plugins/eth2/beacon/test_beacon_node_syncing.py.
- [x] Fix validator test, since the API changed.
- [x] Clean up for unnecessary code, mainly for the tests.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by mhchia** [comment]
Sorry, I forgot there are still todos undone and added reviewers. Please ignore this PR for now and review it later after the todos are solved.

File: eth2/beacon/scripts/run_beacon_nodes.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+dir_root = Path("/tmp/aaaa")
+dir_alice = dir_root / "alice"
+dir_bob = dir_root / "bob"
+port_alice = 30304
+port_bob = 30305
+num_validators = 5
+cmd_gen_testnet_files = f"trinity-beacon testnet --num={num_validators} --network-dir={dir_root}"
+cmd_alice = f"trinity-beacon --port={port_alice} --trinity-root-dir={dir_alice} --beacon-nodekey=6b94ffa2d9b8ee85afb9d7153c463ea22789d3bbc5d961cc4f63a41676883c19 -l debug"  # noqa: E501
+cmd_bob = f"trinity-beacon --port={port_bob} --trinity-root-dir={dir_bob} --beacon-nodekey=f5ad1c57b5a489fc8f21ad0e5a19c1f1a60b8ab357a2100ff7e75f3fa8a4fd2e --bootstrap_nodes=enode://c289557985d885a3f13830a475d649df434099066fbdc840aafac23144f6ecb70d7cc16c186467f273ad7b29707aa15e6a50ec3fde35ae2e69b07b3ddc7a36c7@127.0.0.1:{port_alice} -l debug"  # noqa: E501
```
Comment (ChihChengLiang) [inline eth2/beacon/scripts/run_beacon_nodes.py]: These nodekeys and enode address can be extracted as new variables

File: eth2/beacon/scripts/run_beacon_nodes.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    async def _run(self, cmd: str) -> None:
+        print(f"Spinning up {self.name}")
+        self.proc = await run(cmd)
+        self.tasks.append(asyncio.ensure_future(self._print_logs('stdout', self.proc.stdout)))
+        self.tasks.append(asyncio.ensure_future(self._print_logs('stderr', self.proc.stderr)))
+        try:
+            await self._event_monitor()
+        except EventTimeOutError as e:
+            print(e)
+            self.logger.debug(e)
```
Comment (ChihChengLiang) [inline eth2/beacon/scripts/run_beacon_nodes.py]: Do we need both print and logger.debug?

File: eth2/beacon/scripts/run_beacon_nodes.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                self.logger.debug(f"event \"event.name\" occurred in {from_stream}")
+                self.has_event_happend[event] = True
+
+
+async def main():
+    proc = await run(
+        f"rm -rf {dir_root}"
+    )
+    await proc.wait()
+    # proc = await run(
```
Comment (ChihChengLiang) [inline eth2/beacon/scripts/run_beacon_nodes.py]: remove comment

File: eth2/beacon/scripts/run_beacon_nodes.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    signal.signal(signal.SIGINT, sigint_handler)
+
+    node_alice = Node('Alice\tc2895', cmd_alice)
+    nodes_to_stop.append(node_alice)
+
+    print(f"Sleeping {time_bob_wait_for_alice} seconds to wait until Alice is initialized")
+    await asyncio.sleep(time_bob_wait_for_alice)
+
+    node_bob = Node('Bob\t0e01b', cmd_bob)
+    # node_bob.add_event("stderr", START_SYNCING)
```
Comment (ChihChengLiang) [inline eth2/beacon/scripts/run_beacon_nodes.py]: Unused comment

**Comment by mhchia** [comment]
@ChihChengLiang @NIC619 This should be ready for the review:)

File: tests/core/p2p-proto/bcc/test_full_sync.py [hunk 5]
```diff
@@ -16,6 +16,14 @@
...
 
+class NopBlockImporter:
+    """
+    Do nothing, to override the block validation in ``SyncBlockImporter``.
+    """
+    def import_block(self, block):
+        return None, tuple(), tuple()
+
+
 async def get_sync_setup(request, event_loop, alice_chain_db, bob_chain_db):
```
Comment (ChihChengLiang) [inline tests/core/p2p-proto/bcc/test_full_sync.py]: No Op?
```
NoopBlockImporter
```

Comment (mhchia) [reply tests/core/p2p-proto/bcc/test_full_sync.py]: Ah I derived this name from the instruction in [assembly](https://en.wikipedia.org/wiki/NOP_(code)). I could change it to `Noop` if it is easier to understand.

File: tests/plugins/eth2/beacon/helpers.py [hunk 6]
```diff
@@ -43,11 +38,3 @@
     genesis_block_class=SerenityBeaconBlock,
     genesis_time=0,
```
Comment (NIC619) [inline tests/plugins/eth2/beacon/helpers.py]: Should we set it as `Time.time()`? Or `genesis_time` is not going to be accessed in the tests?

File: eth2/beacon/scripts/run_beacon_nodes.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        return _cmd
+
+    def stop(self) -> None:
+        for task in self.tasks:
+            task.cancel()
+        self.proc.terminate()
+
+    @classmethod
+    def stop_all_nodes(cls) -> None:
+        for node in cls.nodes_to_stop:
```
Comment (NIC619) [inline eth2/beacon/scripts/run_beacon_nodes.py]: Seems like `nodes_to_stop` is not updated anywhere.

File: trinity/plugins/eth2/beacon/plugin.py [hunk 8]
```diff
@@ -74,9 +106,43 @@ def do_start(self) -> None:
...
+        )
+
+        loop = asyncio.get_event_loop()
         asyncio.ensure_future(exit_with_endpoint_and_services(self.event_bus, server))
         asyncio.ensure_future(server.run())
         asyncio.ensure_future(syncer.run())
+        asyncio.ensure_future(slot_ticker.run())
+        asyncio.ensure_future(validator.run())
+        loop.run_forever()
+        loop.close()
```
Comment (ChihChengLiang) [inline trinity/plugins/eth2/beacon/plugin.py]: It looks like after #556 we no longer need these lines.
```
loop.run_forever()
loop.close()
```

File: tests/core/p2p-proto/bcc/test_full_sync.py (hunk 5 above)
Comment (ChihChengLiang) [reply tests/core/p2p-proto/bcc/test_full_sync.py]: A quick search shows that `Noop` is used in the codebase more.

File: trinity/plugins/eth2/beacon/utils.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+override_vector_lengths(XIAO_LONG_BAO_CONFIG)
+
+
+if TYPE_CHECKING:
+    from ruamel.yaml.compat import StreamTextType  # noqa: F401
+
+
+def extract_genesis_state_from_stream(stream: Union[Path, "StreamTextType"]) -> BeaconState:
+    yaml = YAML(typ="unsafe")
+    genesis_json = yaml.load(stream)
```
[the remark this answers is not in the document]
Comment (mhchia) [reply trinity/plugins/eth2/beacon/utils.py]: `yaml.load` raises `ruamel.yaml.scanner.ScannerError` if the content is malformed. IMO it's fine to let the exception raise here.

File: eth2/beacon/scripts/run_beacon_nodes.py (hunk 7 above)
Comment (mhchia) [reply eth2/beacon/scripts/run_beacon_nodes.py]: Good catch!

File: trinity/plugins/eth2/beacon/plugin.py [hunk 10]
```diff
@@ -74,9 +106,43 @@ def do_start(self) -> None:
...
+            validator_privkeys=validator_privkeys,
+            event_bus=self.event_bus,
+            token=server.cancel_token,
+            genesis_epoch=state_machine.config.GENESIS_EPOCH,
+            slots_per_epoch=state_machine.config.SLOTS_PER_EPOCH,
+        )
+
+        slot_ticker = SlotTicker(
+            genesis_slot=state_machine.config.GENESIS_SLOT,
+            genesis_time=state.genesis_time,
```
[the remark this answers is not in the document]
Comment (mhchia) [reply trinity/plugins/eth2/beacon/plugin.py]: I added a field `genesis_time` in `genesis_data`, however it is still initialized from `state.genesis_time`.

File: trinity/plugins/eth2/beacon/utils.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+
+def extract_privkeys_from_dir(dir_path: Path) -> Dict[BLSPubkey, int]:
+    validator_keymap = {}  # pub -> priv
+    for key_file_name in os.listdir(dir_path):
+        key_file_path = dir_path / key_file_name
+        privkey = _read_privkey(key_file_path)
+        validator_keymap[bls.privtopub(privkey)] = privkey
+    if len(validator_keymap) == 0:
+        raise PrivateKeyNotFound("no proper private keys found")
```
Comment (NIC619) [inline trinity/plugins/eth2/beacon/utils.py]: nit: what do you think about `KeyFileNotFound("No validator key file is provided.")`?

**Pull request merged**